### PR TITLE
Use column number in Maven error format

### PR DESCRIPTION
**Pull request by rasmusx**
Hello,

Here is a patch that should fix Maven error format output (#342). Patched worked in local environment running on Linux and with Maven 3.3.9.

### CODE REVIEW

**Comment by brcolow** [comment]
Hmmm...I am currently using Maven 3.3.9 + maven-compiler-plugin 3.3 and I only seem to get line numbers for my errors as you can see here:

`[ERROR] /home/brcolow/code/core/Main.java:[65] q cannot be resolved to a type`

I am not sure why we get different results =(

**Comment by benekastah** [comment]
Given that both results are possible, can we simply include both variations of the errorformat?

**Comment by rasmusx** [comment]
Hm very interesting indeed that you are not getting column number in your output. Idk if it helps but my javac version is 1.8.0_73. Also I had the same issue on OSX env.

Made the new patch there and tested it too. Seems to be working with both outputs now.

**Comment by brcolow** [comment]
I think I figured out why there is a difference. I am actually using the eclipse compiler by using the following configuration for maven-compiler-plugin:

````xml
<plugin>
    <groupId>org.apache.maven.plugins</groupId>
    <artifactId>maven-compiler-plugin</artifactId>
    <version>3.3</version>
    <configuration>
        <source>1.8</source>
        <target>1.8</target>
        <compilerId>eclipse</compilerId>
    </configuration>
    <dependencies>
        <dependency>
            <groupId>org.codehaus.plexus</groupId>
            <artifactId>plexus-compiler-eclipse</artifactId>
            <version>2.5</version>
        </dependency>
    </dependencies>
</plugin>
````

**Comment by rasmusx** [comment]
@brcolow, nice that you found out what caused the difference for us. I think it would be good to have both errorformats still in.

**Comment by blueyed** [comment]
@rasmusx 
Yes.
Can you do this, please?

**Comment by rasmusx** [comment]
@blueyed, yes I could look into it over the weekend.

**Comment by rasmusx** [comment]
Hello again,

Finally I had some time to work on this. I have fixed Maven error format to accept lines in various different formats. Most use cases should be handled.

I have also created one new feature to remove duplicates from job output. I created it due to Maven sending duplicate lines. Could you please take a look at this feature and let me know what you think. It could be useful in future too. Code wise it might not be the best, I don't write vimscript.

Waiting for feedback :)

**Comment by blueyed** [comment]
This branch/PR has conflicts, please rebase it.

**Comment by rasmusx** [comment]
Hi again,

I have rebased the branch with some minor edits. Tests are also passing.
Altough I am not sure why the previous test failed actually. There was just vimlint missing (env setup issue?).

File: autoload/neomake.vim [hunk 1]
```diff
@@ -633,11 +645,62 @@ function! neomake#MakeHandler(job_id, data, event_type) abort
...
+            unlet jobinfo.lines
             let jobinfo.last_register = now
         endif
     elseif a:event_type ==# 'exit'
+        if has_key(jobinfo, 'lines')
+            if maker.remove_duplicates
+                call s:RemoveDuplicates(jobinfo.all_lines, jobinfo.lines)
+            endif
+            call s:RegisterJobOutput(jobinfo, maker, jobinfo.lines)
+        endif
```
Comment (blueyed) [inline autoload/neomake.vim]: This has been changed/removed in ac3cfe8.

File: autoload/neomake.vim [hunk 2]
```diff
@@ -214,7 +215,8 @@ function! neomake#GetMaker(name_or_maker, ...) abort
         \ 'args': [],
         \ 'errorformat': &errorformat,
         \ 'buffer_output': 0,
-        \ 'remove_invalid_entries': 1
+        \ 'remove_invalid_entries': 1,
+        \ 'remove_duplicates': 0
```
Comment (blueyed) [inline autoload/neomake.vim]: Trailing comma here, to make future diffs smaller?!

File: autoload/neomake.vim (hunk 1 above)
Comment (blueyed) [inline autoload/neomake.vim]: The part below also looks like it was not rebased correctly / slipped in?!

**Comment by rasmusx** [comment]
Hi,

Are we going forward with this one? :)

**Comment by blueyed** [comment]
> Are we going forward with this one? :)

Yeah!

Can you please rebase it against master?

File: autoload/neomake.vim [hunk 3]
```diff
@@ -633,8 +645,11 @@ function! neomake#MakeHandler(job_id, data, event_type) abort
                 \ (last_event_type !=# a:event_type ||
                 \  now - jobinfo.start < 1 ||
                 \  now - jobinfo.last_register > 3)
-            call s:RegisterJobOutput(jobinfo, maker, jobinfo.lines[:-2])
-            let jobinfo.lines = jobinfo.lines[-1:]
+            if maker.remove_duplicates
+                call s:RemoveDuplicates(jobinfo.all_lines, jobinfo.lines)
+            endif
+            call s:RegisterJobOutput(jobinfo, maker, jobinfo.lines)
+            unlet jobinfo.lines
```
Comment (blueyed) [inline autoload/neomake.vim]: This looks suspicious: the subscript magic is there for some reason..

Comment (blueyed) [inline autoload/neomake.vim]: Also: where is `all_lines` being filled / added to?

**Comment by rasmusx** [comment]
Hi again!

Sorry for long reply, I was on a vacation :).

Anyway I have rebased it against master once more. Also fixed some function style errors that were not using abort attribute.

**Comment by blueyed** [comment]
I am sorry, but it needs to be rebased again.
Please ping me when done.

**Comment by rasmusx** [comment]
@blueyed , rebased against master with very small change. Please let me know if there is any problems.

File: autoload/neomake.vim [hunk 4]
```diff
@@ -742,6 +744,10 @@ function! s:RegisterJobOutput(jobinfo, lines, source) abort
     let lines = copy(a:lines)
     let maker = a:jobinfo.maker
 
+    if maker.remove_duplicates
+        call s:RemoveDuplicates(a:jobinfo.all_lines, lines)
```
Comment (blueyed) [inline autoload/neomake.vim]: `all_lines` is always empty..?!

**Comment by blueyed** [comment]
Couldn't the removal of duplicate lines also be done via `postprocess`?

**Comment by blueyed** [comment]
There are also quite a lot CI / test failures.

**Comment by rasmusx** [comment]
@blueyed, made new implementation using `postprocess` as you suggested. Everything seems to work, including the tests. Let me know if you have questions.

File: autoload/neomake/makers/mvn.vim [hunk 5]
```diff
@@ -4,6 +4,29 @@ function! neomake#makers#mvn#mvn()
...
+
+    if !has_key(job, 'lines')
+        let job.lines = []
+    endif
+
+    if index(job.lines, text) == -1
+        call add(job.lines, text)
+    else
+        let entry.valid = 0
+    endif
```
Comment (blueyed) [inline autoload/neomake/makers/mvn.vim]: You should not modify `job.lines` here, but instead just keep track of the seen lines yourself.
The postprocessing is meant to massage the location/quickfix list only.

See https://github.com/neomake/neomake/issues/766, where I thought about passing `jobinfo` into the `postprocess` function, which you could use then - and we would not have to change the `entry` temporarily.

**Comment by blueyed** [comment]
@rasmusx 
ping

**Comment by deathbeam** [comment]
I am using maven-compiler-plugin 3.5.1 and I can confirm that it is too also reporting line and column numbers in format [%l\\,%c]

**Comment by blueyed** [comment]
@deathbeam 
Thanks for the confirmation, but for me the actionable on this PR is still unclear.

File: autoload/neomake/makers/mvn.vim [hunk 6]
```diff
@@ -4,6 +4,29 @@ function! neomake#makers#mvn#mvn()
     return {
          \ 'exe': 'mvn',
          \ 'args': ['install'],
-            \ 'errorformat': '[%tRROR]\ %f:[%l]\ %m,%-G%.%#'
+         \ 'remove_invalid_entries' : 1,
+         \ 'postprocess': function('neomake#makers#mvn#InvalidDuplicates'),
+         \ 'errorformat':
+           \ '%E[%tRROR]\ %f:[%l]\ %m,' .
+           \ '%E[%tRROR]\ %f:[%l\,%v]\ %m,' .
```
Comment (deathbeam) [inline autoload/neomake/makers/mvn.vim]: Just wondering, why %v and not %c? I think that maven-compiler-plugin is reporting tabs as single column, or am I wrong?

**Comment by dpc** [comment]
Any love for scala? Also any way to easily use `mvn compile` instead of `mvn install`?

The mvn version I work with uses:

```
[ERROR] /home/dpc/lab/myproject/src/java/Somefile.scala:1401: error: illegal start of declaration (possible cause: missing `=' in front of current method body)
[ERROR]     return ""
[ERROR]   
````

**Comment by blueyed** [comment]
@rasmusx 
I hope you've not given up on it?

Would anybody like to pick it up and finish it otherwise?

**Comment by blueyed** [comment]
No update since a while, so I am closing this.
Feel free to pick this up though, of course (anybody!).

**Pull request closed without merging**